homeassistant/components/lutron_caseta/__init__.py

```diff
@@ -375,19 +375,25 @@ def _handle_none_keypad_serial(keypad_device: dict, bridge_serial: int) -> str:
     return keypad_device["serial"] or f"{bridge_serial}_{keypad_device['device_id']}"
 
 
-def _area_name_from_id(areas: dict[str, dict], area_id: str) -> str:
+def _area_name_from_id(
+    areas: dict[str, dict], area_id: str, previous_area_part: str = ""
+) -> str:
     """Return the full area name including parent(s)."""
 
     if area_id is None:
         return UNASSIGNED_AREA
 
     area = areas[area_id]
-    if "parent_id" in area:
-        parent_area = area["parent_id"]
-        if parent_area is not None:
-            return f"{_area_name_from_id(areas, parent_area)} {area['name']}"
+    if (parent_area := area["parent_id"]) is None:
+        # This is the root area, return last area
+        return previous_area_part
 
-    return area["name"]
+    if previous_area_part:
+        return _area_name_from_id(
+            areas, parent_area, f"{area['name']} {previous_area_part}"
+        )
+
+    return _area_name_from_id(areas, parent_area, area["name"])
```

Comment:
Can you rewrite this recursion so it doesn't do further down the stack only to return right away when it hits the `if (parent_area := area["parent_id"]) is None:` conditional?

Reply:
Untested but I think you could write it has

```python
def _area_name_from_id(areas: dict[str, dict], area_id: str | None) -> str:
    """Return the full area name including parent(s)."""
    if area_id is None:
        return UNASSIGNED_AREA
    return _construct_area_name_from_id(areas, area_id, [])


def _construct_area_name_from_id(
    areas: dict[str, dict], area_id: str | None, labels: list[str]
) -> str:
    """Recursively construct the full area name including parent(s)."""
    area = areas[area_id]
    parent_area_id = area["parent_id"]
    if parent_area_id is None:
        # This is the root area, return last area
        return " ".join(labels)

    labels.insert(0, area["name"])
    return _construct_area_name_from_id(areas, parent_area_id, labels)
```

Reply:
Also since we do this in multiple places, I'd decorate `_area_name_from_id` with `@lru_cache` since recursion is inherently expensive in python

Reply:
Fails with the @lru_cache decorator.  Maybe I'm not using it correctly.

Reply:
otherwise works well

Reply:
in _construct_area_name_from_id(
had to change  area_id: str | None    
to area_id: str

Reply:
I've included these changes, other than the @lru_cache decorator

Reply:
The dict is not going to be hash-able so you'd have to make it into an object to with its own cache to be able to cache it.

That can happen later though